### ISSUE TITLE
LokiJS Serialization Overhaul (Opt-In)

### ISSUE THREAD

**Issue by obeliskos** [techfort/LokiJS#498]
JSON serialization (stringify/parse) seem to have an exponential amount of overhead when dealing with large (deep) object serialization.  This 'Issue' is more of documentation and rfc for refactoring for loki persistence and serialization as I expect to make this my next priority.

It is a little challenging to diagnose memory issues with node, so I created a stress.js node script under benchmarks to help diagnose this.  Unfortunately it seems my mere act of randomly generating 40000 or so objects to get a 60Meg database ends up using up several gigs of heap space before even getting to serialization or persistence.  Invoking gc did not seem to help.

Initial estimate is that if you do not expect your database to exceed 40-50 megs then you will not likely encounter any issues.  If you expect your process space to have less than 1 or 2 gigs available, you may hit whatever stack/heap limitation I encountered using the apparently memory expensive JSON serialization.

My initial thoughts are to possibly re-architect loki serialization be driven off of a 'serializationMethod' (name undecided) which might be of any of the following values (passed in on loki construction) : 
- "prettyJSON" : will output larger but formatted for human reading (least efficient)
- "normal" (default) : dumped as 1 big line as part of 1 JSON.stringify() call
- "destructure" : 
  - database and all collections but no data dumped as 1 big line as part of 1 JSON.stringify call
  - each collection data object iterated and stringified individually and dumped as 1 line within same file.
  - likely would need a collection.length property to be serialized in first step/line to know when next collection starts.
- "partitioned" : 
  - similar to "destructure" except database (container) is in own file and each collection's data dump will go into different files
  - if our existing adapters are key/value where you pass them dbname/json, they will receive "dbname", "some json"  and "dbname-users", "some string combination of many json dumps"
  - This approach may not only allow larger databases but reduce time it takes to save large database where only 1 collection has been modified.

Is still at an early stage so these numbers are still pure conjecture but I would guess when done, I myself would use prettyJSON for databases under 10megs or so, normal for databases 10-40 megs, destructure for databases 40-100megs, and maybe partitioned for 100+ meg databases.  I guess it depends on the speed of your file system and how much ram you have.  also 'prettyJSON' files will be larger than 'normal' files due to the addition of whitespace and formatting, so if you never expect to read it, you would leave at normal.

There may be other areas where we enhance our 'reference' mode persistence adapters to save using methods other than serialization either through binary or passing objects rather than serialized strings to indexed db to see if this has improved effect.  Persistence adapters not included in loki distribution may employ whatever dependencies they wish so optimal alternatives may be pursued for high end use.

I will likely be testing and/or phasing in such an option and modes to gauge and reduce the memory overhead of serialization across the board for loki.  Such refactoring and testing should take a while.  I do not expect to break compatibility with this and I am not sure if I will implement this first on our 2.0.0 branch, but I expect this will likely be get put into main 1.4.x version branch and npm as well.  If anyone has opinions or interest in collaborating, feel free to chime in.

**Comment by techfort** [comment]
hey @obeliskos long time no speak :) ! Sorry for the absence, i mentioned the crazy commute right? Anyway, very interesting post! Have you looked into msgpack? there's even a loki msgpack adapter that seems promising. The whole serialization overhaul definitely belongs to v2.0.0 as I assume that one way or another it will majorly break the API for persistence.
Another way I can think of to reduce space and load times is the approach i was taking in that petty project of mine http://github.com/techfort/jumble : this would only work with fixed-structure objects, which should really be the case for most uses. 
Also, some food for thought: lack of time has halted development of the incremental adapter I started a couple of months back, however I made a start on something that would probably fit into the "partitioned" serializer, as the database metadata is in a root folder, collections are folders, and documents are single, append-only files, where all version of a document are stored in an array and the version at the last index is the current version.
Time is still not my friend but I'd be very interesting in working on this.

>

**Comment by obeliskos** [comment]
Thanks techfort, completely understand... I have only briefly looked at some third party libraries.  Most are geared towards -producing- small end product for transmission purposes.  I think loki would be interested in / prioritize the serialization method with the least memory overhead.  Some testing in that area would be useful with process.memoryUsage() diagnostics.  I will take a look at your projects and see if I get ideas for feedback next time I have a few moments.

I need at least option of normal and prettyJSON in 1.4, my thinking was we could have reusable logic for all adapters or pass of to reference mode adapters for no (loki) serialization needs.  But I am in no position for a quick implementation of last two so maybe warm up / test on a specific adapter would be best place to start.

**Comment by obeliskos** [comment]
Just an update, i updated my diagnostic scripts to use a more memory efficient algorithm.  The result is still a significant overhead when calling db.serialize() but the overhead of my old random object generation should be skewing results much less now.  Node can easily handle 100mb database of 130,000 objects without raising memory allocation and (chrome) browser can handle about the same.

Here's a browser test if you don't mind it temporarily using up about 100megs of indexed db storage (should be able to delete in developer tools) : 
[Browser Stress Test](https://rawgit.com/techfort/LokiJS/master/examples/sandbox/LokiSandbox.htm#rawgist=https://gist.githubusercontent.com/obeliskos/3d4884a52f036288553f2704b65b4505/raw/7e245b0cef0015c68302e51156963bd965352fe9/idb%2520stress)

benchmarks/stress.js now tests just creating, serializing and saving and a new benchmarks/destress.js now tests loading, and then reserializing so you can see deserialize overhead and then serialize overhead.  Need to run stress.js to create db before you can run destress.js.

**Comment by obeliskos** [comment]
I added 'serializationMethod' loki option which now supports 'normal' and 'pretty' values (use pretty only if you want to easily read the serialized JSON. 

I also added an experimental 'serializationMethod' of 'destructured'.  This data format may be tweaked so it is not intended for production use.   Currently only small benefits are gained from this method because I need to keep all 'lines' of destructured JSON in memory and then join them for 3x memory overhead.  But I did notice an increase from 340K documents to around 390K documents gained from lower memory pressure of not having to deep scan serialize in a single operation.   To support 'destructured' serializationMethod, I added a new loki 'destructureDelimiter' option as well (which has a default value of '$<\n')

So for now use as 'normal' or 'pretty' if you need to.  Still early stages, will probably work on a 'reference mode' filesystem adapter as a standalone module/adapter next to lower memory pressure via file streaming each line out as I generate them rather than aggregating and joining into single string.

**Comment by obeliskos** [comment]
While still 'experimental' and a 'work-in-progress', I have expanded destructuring options to output a wider variety of formats to use as a building block for custom adapters or synchronization.

So essentially in destructuring we are disassembling the database into individual document serializations along with an empty database serialization and reassembling it into one of these four formats :

- D : (Default) one big Delimited string { partitioned: false, delimited : true }
- DA : Delimited Array of strings [0] db [1] collection [n] collection { partitioned: true, delimited: true }
- NDA : Non-Delimited Array : one iterable array with empty string collection partitions { partitioned: false, delimited: false }
- NDAA : Non-Delimited Array with subArrays. db at [0] and collection subarrays at [n] { partitioned: true, delimited : false }

Additionally for partitioned formats (DA, NDAA) you can optionally pass a 'partition' option of -1 to receive only the db container (empty of data), or 0-n to serialize only a specific collection.  If partition option is not passed, it will return the db and all collections in their respective (internally partitioned) formats. 

Most of this functionality is to use as building blocks for custom persistence adapters which either emphasise low memory overhead (bigger databases) and/or lower disk IO times by not having 
to save unmodified collections whenever one collection needs saving.

I will next look at how complete i can make a generic 'deserializeDestructured' method to inflate from these formats. The main uncertainty will be deserializing individual partitioned collections which will likely be driven from our async persistence adapter model.  That functionality may be left to individual adapters or i will create recursive async iterator method to deserialize each collection only 'one at a time' rather than simulatenously (thus adding memory overhead).

Once deserialization of those formats is done, I will unit test and look at using them within an optimized loki filesystem adapter, seeing where improvements are notable and possibly tweaking.  (I did say this would take a while).  Improvements beyond that point might be a more granular streamable approach which may require the use of es6 generators to yield individual lines to a line-based serializable reference mode adapter using something like node fs (rather than key-value pair).

I would like to think the current four destructured formats are mostly final, and they probably will be pending any unanticipated issues deserializing from them.

**Comment by techfort** [comment]
hey @obeliskos - i'm trying to set some time aside to go through this in
detail. It looks great, but i want to understand exactly what is going on.
Thanks for the great work in the meantime.

**Comment by obeliskos** [comment]
Yea if you get time we can cooperate.  I've been a little slow adding pieces to this but I just checked in a deserialization component.  So as I mentioned I am supporting four formats.  If you want to see the output, I made a loki sandbox gist [here](https://rawgit.com/techfort/LokiJS/master/examples/sandbox/LokiSandbox.htm#rawgist=https://gist.githubusercontent.com/obeliskos/98a73205d7fe9746a687634e19a5eb89/raw/3821c9bbb6bae2689b00d16be9fae78dff430e28/destructuring%2520demo).

So if you load that up, by default it will output a large array of strings good for line-by-line serialization which is NDA format.

Formats are somewhat explained in gist and I will eventually write up in wiki, but another good format might be 'DA' format which would be simple array containing 1 string for db, and 1 string for each collection.  These could be sent off to different files.  To preview what that would look like make options look like this for both serialization and deserialization calls : 

```javascript
{
  partitioned: true,
  delimited: true
}
```

For a custom loki fs adapter I was thinking about using an NDAA format using these options : 
```javascript
  partitioned: true,
  delimited: false
```
and sending db off to one file, then iterate remaining array items and stream that subarray, one at a time.

**Comment by obeliskos** [comment]
I have borrowed from the logic created above to create a new loki-fs-structured-adapter for use with nodejs.  Since it is able to tap into node streams, I adapted the destructure logic to use es6 generator/iterators to 'stream' the database to and from disk as a loki 'reference' persistence adapter.

Using the default node memory allocation, before I was limited to about a 175meg database.  Initial results now (with a single huge collection) are at least 400meg databases are possible.  Since each collection is streamed to its own file, that limit may be exceeded if you either : 
- boost node memory allocation, or
- have multiple collections under 400 megs or so each

Given adequate disk speeds, memory allocations, and partioning of data over multiple collections, this should be as much scaling as needed for near future.

I have only done initial testing (eyeballing output files, verifying collection counts, and making sure file sizes seem appropriate).  My stress.js and distress.js scripts in the benchmark folder can be used to duplicate my tests.

If you don't already have a node debugger and you are running on node (6.3+ ?) they added a nice  [Chrome DevTools debugging capability](https://medium.com/@paul_irish/debugging-node-js-nightlies-with-chrome-devtools-7c4a1b95ae27) which was very useful for this task.

At some point I may create additional indexdb adapter at least for partitioning capability.  Near term I will likely test this new adapter more until I feel comfortable and then I'll update wiki.

**Comment by techfort** [comment]
This sounds amazing Dave, I really hope to get a minute of time this
weekend to explore this properly!!

**Comment by obeliskos** [comment]
Ok I think I am rounding out the bulk of changes I wanted to make.  

So to cover non-node environments I have settled at least for the time period to just build a higher level wrapper adapter which converts existing non-reference mode adapters (essentially key/value stores) into a partitioned 'reference mode' database adapter.  

I also have added a new LokiMemoryAdapter for diagnostic purposes.  It is essentially a key/value store implemented with internal save counts and last save functionality.  This is volatile memory storage not intended for actual persistence purposes but allows me at least some level of diagnostics. I should also be able to use this LokiMemoryAdapter as a mock persistence adapter for unit testing.

To see a demo of these two new adapters used together for diagnostic purposes, I have created this [Loki Sandbox gist](https://rawgit.com/techfort/LokiJS/master/examples/sandbox/LokiSandbox.htm#rawgist=https://gist.githubusercontent.com/obeliskos/15c1aa87da16cd89b328eb84bbcdf8fa/raw/d91ac3fee212dc5aa96cb05f479d825faa17c1c8/PartitionedMemoryAdapterTest) which you can run.

Since this functionality is opt-in I will begin documenting as experimental and continue testing myself.  If anyone finds any bugs, posting here would be advisable.  Also my es6 and node streams is not yet up to speed so any improvements are spotted, feel free to pass those on as well.

**Comment by obeliskos** [comment]
Ok while I am sure I have a lot of areas for improvement I now feel the diagnostic stress and destress scripts in the benchmark folder are accurate.  I fixed a reliability issue for random string generation (for random object properties) by using node crypto library.

Now I have a more 'clean' environment, which I would guess is (still?) 1.4gb because most of my tests fail when I push heap usage up to around 1.45gb or so.  ***If your javascript is already using up lots of memory loki will need to compete for that memory***.  For node you can boost heap allocation via command line switches.

So here are my results for node, all but the new node adapter will probably translate to web : 

**For 'One Big Collection' :**
'normal' and 'destructured' serialization methods ran into limit around 250mb.  I saw no significant reduction in memory overhead but what I did not expect was that the save was faster (~7.2s -> 5.48s).  Many separate JSON serializations appears to be a lot faster on cpu than one big one.

Partitioning Adapter not able to take advantage when only one big collection... limit was ~250mb @5.5s to save.

LokiFsStructuredAdapter : The 'streaming' nature of the adapter enabled me to create ~940mb database... at that size it took 19s to save though.  The closer you get to heap limit the slower the io due to smaller buffer?

**For 'Two Equally Big Collections' :** 
'normal' and 'destructured' : database limit still 250mb ... pretty much have to have both collections add up to same limit.

Partitioned Adapter : This actually helped a lot... and database was able to grow to ~237mb * 2 for total of ~474mb database.  Save times in the high 4s range.

LokiFsStructuredAdapter : 
Each collection was able to grow to ~474mb, which * 2 collections is ~950mb database.  Considering the heap allocation is around 1.4gb that's pretty great.  The 17s is pretty high... but maybe if I left more free heap space the buffer would allow faster saves.

**So things I learned :** 
- 'destructured' serializationMethod seems to speed up database saves when using non-reference mode adapters like original LokiFsAdapter and IndexedAdapter. 
- Partitioned adapter seems very useful with benefits for save speeds as well as total database size.  If your database has memory usage distributed among your collections then serializing the whole db in these smaller chunks lowers memory pressure on all environments.
- LokiFsStructuredAdapter seems best if you intend to push limits.  On equally sized databases it leaves more free heap space for your program, or if you need bigger databases that memory efficiency will allow larger database saves.  Compared to the Partitioning adapter, you don't need to consider individual collection sizes just whole database size when determining if you will exceed heap usage.

**Comment by obeliskos** [comment]
So Chrome was the outlier use case which has a quirky limit on how big a single indexed db 'row' can be.  I discovered this in #484 and that sent me on this journey to stress memory limits. 

In Chrome, the database can be bigger but its limits the size of individual records.  So I added the ability to do paging to the LokiPartitioningAdapter.  So now it can not only break up each collection into its own file but if collections exceed a certain amount of memory to persist, it will break it up into pages.

Chrome seems limited to around 60megs but that may be Unicode so I made the default page size 25 megs.  This is pageSize configurable with options.

Here is an example [stress test loki sandbox gist](https://rawgit.com/techfort/LokiJS/master/examples/sandbox/LokiSandbox.htm#rawgist=https://gist.githubusercontent.com/obeliskos/2ed332f469a39c94fe1c17d1251234b5/raw/ca332bea67eaa0296933175fd024b70b0d3b821a/Chrome%2520IndexedDB%2520Stress%2520Test) if you want to try out or see how to use this paging feature.

I have also updated the [Persistence and Adapters](https://github.com/techfort/LokiJS/wiki/LokiJS-persistence-and-adapters) wiki page to more cleanly present the persistence arena as it now exists.
